File: src/Java/algorithms/treeandgraph/BinaryTreeLevelOrderTraversal.java
package algorithms.treeandgraph; 

import java.util.*; 

public class BinaryTreeLevelOrderTraversal {
    /* Function: Traverse the binary tree in level order. 
     * Solution: Use the queue size as the counter. 
     */ 
    
    public List<List<Integer>> levelOrder(TreeNode root) {
        List<List<Integer>> res = new ArrayList<List<Integer>>(); 
        if (root == null) {
            return res; 
        }
        
        Queue<TreeNode> queue = new LinkedList<TreeNode>(); 
        queue.offer(root); 
        while (!queue.isEmpty()) {
            int size = queue.size(); 
            List<Integer> list = new ArrayList<Integer>(); 
            for (int i = 0; i < size; ++i) {
                TreeNode cur = queue.poll(); 
                list.add(cur.val); 
                
                if (cur.left != null) {
                    queue.offer(cur.left); 
                }
                if (cur.right != null) {
                    queue.offer(cur.right); 
                }
            }
            res.add(list); 
        }
        
        return res; 
    }
} 

File: src/CPP/src/leetcode/BinaryTreeLevelOrderTraversal.cpp
/*
 * [Source] https://leetcode.com/problems/binary-tree-level-order-traversal/
 * [Difficulty]: Easy
 * [Tag]: Tree
 * [Tag]: Breadth-first Search
 */

#include <iostream>

using namespace std;

// [Solution]: Recursive call with depth.
// [Corner Case]:
class Solution {
};

/* Java solution
public class Solution {
    public List<List<Integer>> levelOrder(TreeNode root) {
		List<List<Integer>> result=new ArrayList<List<Integer>>();
		if(root==null) return result;
		levelOrder(root, 0, result);
		return result;
    }
	
	public void levelOrder(TreeNode root, int level, List<List<Integer>> result){
		if(root==null) return;
		if(result.size()<=level){
			List<Integer> list=new ArrayList<Integer>();
			list.add(root.val);
			result.add(list);
		}else{
			result.get(level).add(root.val);
		}
		levelOrder(root.left, level+1, result);
		levelOrder(root.right, level+1, result);
    }
}
 */

// [Solution]: Use a queue to do breadth first search. Use size to track the current level is end or not
/* Java solution
https://github.com/ZhenyingZhu/ClassicAlgorithms/blob/master/src/algorithms/treeandgraph/BinaryTreeLevelOrderTraversal.java
 */

int main() {
    Solution sol;

    return 0;
}
